### ISSUE TITLE
Estimate and plan for - Deploy HL to Production Environment

### ISSUE THREAD

**Issue by katyedwards** [healthlocker/healthlocker#1061]
- [ ] T6.5h - 7h for setting up the environment
  - [ ] Add necessary languages/packages (t2h)
  - [ ] Buffer because installing ODBC has been problematic in the past (t2h)
  - [ ] Create environment variables (t25m)
  - [ ] Write up documentation to ensure SLaM has the correct info for the server configuration (will be based off of [this documentation](https://github.com/healthlocker/oxleas-adhd/blob/deploy_ssl_docs/deployment_and_ssl_doc.md) and updated where needed  to account for ePJS etc)  (T2h - 2.5h)
- [ ] T2h for deployment
- [ ] T25m Create database and tables on server
- [ ] T2h for ssl
- [ ] T25m Update Heroku DNS settings (to allow us to use healthlocker.uk)
- [ ] T25m configure SSL auto-renew, see #1063
- [ ] T1h for nginx configuration for the new environment
+ [ ] T1h for standups and catch ups throughout
+ [ ] T4h reserved for potential additional debugging (may not be required)

**Comment by reddog** [comment]
@katbow Thanks very much for your list of setup items above - really helpful to see that in one place.

I've modified to include SSL auto-renew setup, see #1063

**Comment by reddog** [comment]
@iteles Thanks for your breakdown and time estimates.  Can you confirm if these have been given in view of yesterday's deployment of Oxleas to live.  Starting from a fresh server the process was complete within one developer day.

See https://github.com/healthlocker/oxleas-adhd/issues/168 and https://github.com/healthlocker/oxleas-adhd/issues/187

Can you please clarify what will be the output of

> Write up documentation to ensure SLaM has the correct info for the server configuration

will be.  Can this be a further enhancement/refinement of https://github.com/healthlocker/oxleas-adhd/issues/191 ?  In which case, does the time estimate need updating?

Can you please clarify why

> spike to investigate deploying from a branch, to be used for estimating how long it would be to implement this.

is required.  What will this enable?  This has been mentioned elsewhere and would appear to be a larger development team learning as opposed to a project-specific requirement.

Is

> Update Vagrant file. Other devs cannot run this app easily, so need to investigate the problem with vagrant, and update the configuration as we set up the server.

part of the deploy to production process?  Is this required for deployment?  If not, can it be separated out to its own issue.

**Comment by iteles** [comment]
@katbow Happy Monday! Assigning these questions to you as you analysed, broke down and estimated the original tasks ❤️

**Comment by reddog** [comment]
@iteles Thanks for passing this on.  Do we have an estimate on when revised timings will be able available?

**Comment by reddog** [comment]
In order to both test and refine the documentation for Oxleas (making it SLaM HealthLocker specific), can the documentation at https://github.com/healthlocker/oxleas-adhd/issues/191 be used as a guide for this installation?  Would there be any problems in this?

**Comment by reddog** [comment]
@iteles Although not yet complete, Is the documentation at https://github.com/healthlocker/oxleas-adhd/issues/191 at a point that could enable better time estimating of this issue?

Can you please confirm when the estimates on this issue will be reviewed.  That should happen before Thursday when work is currently expected to begin.

**Comment by reddog** [comment]
@RobStallion Thanks for further refining the list of tasks required for this issue and the time estimates.

Are you able to reply to my questions in https://github.com/healthlocker/healthlocker/issues/1061#issuecomment-328065421 ?  Some items have been removed altogether, are they no longer required or have new issues been opened?

The new time estimates put this release at about 2.5 days. Is that paired days?  If so this issue would be estimated at  about 5 developer days. The release you just completed for Oxleas was about 1.5 of just your time.

Can you expand further on why this release requires the extra roughly 3.5 person days? (assuming timing here are for paired work)

**Comment by iteles** [comment]
@reddog Have just been reviewing this ahead of tomorrow. 

Apologies that this issue wasn't made very clear! When the re-estimating was done, this was done as a single developer in light of the Oxleas work carried out, so this is in fact 2.5 days for one developer 👍  The complexities of the ePJS connection and requirement for ODBC make this  a more complex deployment than Oxleas.

With regards to your questions above (https://github.com/healthlocker/healthlocker/issues/1061#issuecomment-328065421):
+ The plan here, as you suggested above, is to use the documentation created for Oxleas (@robstallion has reflected this the re-estimation write-up above) - we know that this documentation was cut down to be more basic than we had originally intended, so it is likely that as we go through it now, changes will need to be made as we go through to ensure it makes sense as stand alone documentation to new developers; this is what is intended here
+ The spike around deployment from a branch - my understanding is that this was discussed during the previous Healthlocker sprint and at that point, was understood to be a requirement of your for a forthcoming sprint; from your line of questioning it seems that this is not the case so it has been removed (though we're happy to discuss this tomorrow of course)
+ Updating the vagrant file is a recommendation to facilitate work for a diverse set of potential developers (both in the NHS and out) - I agree we can have this as a separate issue, which we discuss on its own 👍

**Comment by reddog** [comment]
@iteles Thanks very much for your reply and clarifying further the re-estimates.  Thanks also for clearing up that these are single developer estimates, not the usual paired developer estimates.

Deployment of Oxleas *and* documentation was complete in about 1.5 developer days - without deployment documentation existing in one place in advance of the task.  The documentation that @RobStallion has completed at https://github.com/healthlocker/oxleas-adhd/issues/191 would seem to do a great job of pulling together notes spread across various issues plus including some details not mentioned elsewhere at all.

Performing deployment *with* the aid of having a single documentation to follow should further reduce the time required (from Oxleas' 1.5 days).  I would be interested to follow if that is or is not the case in this deployment so as to better understand any areas where we need better understanding of the issues contained.

ODBC setup is not included in the Oxleas document - estimates above put this at about 2 hours work.

Can a copy of the Oxleas document be brought in to this repo and tailored and improved as required.  Any iterative improvements to the document would be great to have captured in this version.

>  this documentation was cut down to be more basic than we had originally intended, so it is likely that as we go through it now, changes will need to be made as we go through to ensure it makes sense as stand alone documentation to new developers

I share your desire for more effective documentation, which it would seem @RobStallion's work is.  Being a first version (as was requested), this should be continued to be built upon.  As per the document for Oxleas' deployment, the acceptance criteria for the document for deploying this repo should be that the repo and deployment document can be taken and by and largely followed in a step by step process to complete a deployment to a fresh Linux VM (of which the Azure instances we have are).

> spike around deployment from a branch

I am not sure if I specified this previously - I haven't been performing deployments.  I think it has been mentioned in other projects.  If this is a wider development team issue than I'm not sure if it belongs as part of this project.  Of course, if it is something that's needed specifically for this project then an issue should be opened for it.

> Updating the vagrant file is a recommendation

As suggested, please do open an issue for this detailing what we currently know about issues with the Vagrant file and what impact not fixing that would have.

**Comment by iteles** [comment]
@reddog Thanks for your questions on this. 
Some thoughts on this which we can discuss on our kick off call this morning:
+ The oxleas deployment consisted of just over 2.5 days:
  + Configuration + deployment (1 day)
  + Documentation (around half a day, 2 initially plus additional details requested)
  + SSL nginx config (1 day) https://github.com/healthlocker/oxleas-adhd/issues/94
  + Configuring let's encrypt to auto-renew https://github.com/healthlocker/oxleas-adhd/issues/193
+ The Healthlocker deployment above has been estimated at about 1.75 days for a more complex deployment (including the ePJS component)
  + There is also another 4 hours of debugging buffer (which I believe is a sensible precaution given the additional complexity) as well as accounting for 15mins for each of 4 calls (kick off + 3 standups)

I hope this clarifies that the work done for Oxleas has in fact contributed to significantly reducing the amount of time required for the Healthlocker deployment.  

In terms of documentation, I think we are very aligned - though we won't reach perfect documentation in the 2 hours allotted, we will have made iterative improvements. 
My recommendation would be for a canonical version of the documentation to live in one repo with deployment specifics in further repos so as to reduce the maintenance burden for these.

**Comment by reddog** [comment]
Moved actual deployment tasks to #1067 in order to keep things a little tidier.  This issue has become a useful conversation on time estimating and planning, so now split out the doing part.
